### PR TITLE
Fix and test autograd with hidden units

### PR DESCRIPTION
**Pull request by kevinchern**
- Fix the autograd computation when hidden units are present. The gradients were computed incorrectly because the linear and quadratic parameters were used to marginalize hidden units without detaching from the computation graph. The fix is to detach them when computing effective fields.
- Add a test to check the gradient is as expected.

Thanks @mhramani for flagging the lack of tests, which lead to discovering this bug!

### CODE REVIEW

File: tests/test_boltzmann_machine.py [hunk 1]
```diff
@@ -356,6 +356,46 @@ def test_objective(self):
...
+        #                v3 <0.3>
+        s_observed = torch.tensor([[1.0, -1.0]])
+        s_model = torch.tensor([[1.0, -1.0, 1.0]])
+        quasi = grbm.quasi_objective(s_observed, s_model, "exact-disc")
+        quasi.backward()
+        # Compute gradients manually
+        # Compute unnormalized density
+        #                            h1v1 + h2v2 + h3v3 + J23v2v3 + J12v1v2 + J13v1v3
+        q_plus = torch.exp(-torch.tensor(0.2 + 0.2 - 0.3 - 0.6 + 0.2 - 0.3))
+        q_minus = torch.exp(-torch.tensor(-0.2 - 0.2 + 0.3 - 0.6 + 0.2 - 0.3))
```
Comment (mhramani) [inline tests/test_boltzmann_machine.py:381]: looks like this is not sorted  according to `h1v1 + h2v2 + h3v3 + J23v2v3 + J12v1v2 + J13v1v3`:
I think it should be
 q_plus = torch.exp(-torch.tensor(0.2  +  0.2  - 0.3  - 0.6 + 0.2 - 0.3))
 q_minus = torch.exp(-torch.tensor(-0.2 + 0.2 - 0.3 - 0.6 - 0.2 + 0.3))
 this won't change the results

**Review by mhramani — APPROVED** [review]
LGTM, made a minor comment

File: tests/test_boltzmann_machine.py (hunk 1 above)
Comment (kevinchern) [reply tests/test_boltzmann_machine.py:381]: great catch, thanks!

**Comment by mhramani** [comment]
@kevinchern this is a very important and time-sensitive PR,  I already reviewed approved, could you please merge.  @thisac Please have a look once you get a chance.

**Comment by kevinchern** [comment]
Thanks @mhramani
